### ISSUE TITLE
Occasionally seeing duplicate messages

### ISSUE THREAD

**Issue by wavexx** [terminal-discord/weechat-discord#80]
I occasionally had the following behavior:

- connect using "/discord connect"
- switch to a DM buffer with someone
- type some message
- see my own messages appear twice in the buffer

In these cases I've found that just reconnecting discord wouldn't fix the problem. I had to quit weechat and restart.

I'm using weechat 2.8 (from debian unstable) and rebuilt weechat-discord from master  at 508b3ac94f98b2fa17dfe549c0c644ab3fe67c78

**Comment by cuppajoeman** [comment]
I had an issue similar to this, but I believe it was because I tried to connect while I was already connected. I disconnected once and sent a message and only one instance of my message appeared in my buffer.

**Comment by wavexx** [comment]
Mmh, I can confirm I can connect twice and see duplicates.
Shouldn't this be checked-for?

**Comment by Noskcaj19** [comment]
Mmm, you should not be able to reconnect after you have connected to discord, however, the code only prevents reconnects once you have connected, it will not stop you from reconnecting until discord has sent it's Ready event.

**Comment by wavexx** [comment]
I had this happen frequently when loosing connectivity.

Could it be that multiple attempts to reconnect can start concurrently resulting in multiple connections?

**Comment by Noskcaj19** [comment]
I suppose it might be possible for multiple "Ready"s to be received.  However, history should only be loaded the first time a buffer is viewed.

It might be helpful to know if the duplicates remain after running `/discord rehistory`.

**Comment by wavexx** [comment]
Note that I'm seeing duplicates of messages of my _own_ messages just as I send them. I suspect rehistory wouldn't fix that.

**Comment by wavexx** [comment]
So I just rebuilt from master. Connect with /discord connect. Switch do a DM buffer.
Waited some seconds. Connected with /discord connect again.

I'm now seeing duplicates of both the messages I send and of what I receive from the DM.

/discord rehistory clears the buffer and fixes the old messages, but duplicates of new messages still appear.
Looks like multiple connections are running in parallel.

**Comment by Noskcaj19** [comment]
Rehistory should reload the entire buffer contents.  Personal messages are not echoed locally, they (should) only appear after round tripping to Discord.

**Comment by Noskcaj19** [comment]
That is very interesting... I'm not actually sure how multiple connections would work, since a global variable is used to manage the connection.

**Comment by wavexx** [comment]
I suppose it doesn't work? ;)
I just tried to run /discord connect again, and sure enough, I get 3 copies of everything.

**Comment by Noskcaj19** [comment]
Very interesting.  Is the plugin printing `Discord connected` to the main buffer?

**Comment by wavexx** [comment]
Yes. Interestingly, it prints "Already connected" *first*, immediately followed by a new "discord connected".

```
16:00:17    weechat    | discord: Discord connected
16:15:31    weechat    | discord: Already connected
16:15:33    weechat    | discord: Discord connected
```

**Comment by Noskcaj19** [comment]
Oops.  Just double checked the code.  It prints the "already connected" warning, but doesn't actually return...

**Comment by wavexx** [comment]
Looks good to me now, I'll close this.
